README.md
[![Build Status](http://13.238.107.244:8080/job/Empirical_Engine/badge/icon?build=last:${params.ghprbActualCommit=master})](http://13.238.107.244:8080/job/Empirical_Engine)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# Empirical_Engine
Contains codes to calculate Empirical IMs

To calculate empirical IMs:

```
usage: calculate_empirical.py [-h] [--vs30_file VS30_FILE]
                              [--vs30_default VS30_DEFAULT]
                              [-r RUPTURE_DISTANCE] [-srf SRF_INFO]
                              [-s STATIONS [STATIONS ...]]
                              [-rm MAX_RUPTURE_DISTANCE] [-i IDENTIFIER]
                              [-c CONFIG] [-e] [-p PERIOD [PERIOD ...]]
                              [-m IM [IM ...]]
                              output

positional arguments:
  output                output directory

optional arguments:
  -h, --help            show this help message and exit
  --vs30_file VS30_FILE, -v VS30_FILE
                        vs30 file. Default value is 250 if station/file not
                        present
  --vs30_default VS30_DEFAULT
                        Sets the default value for the vs30
  -r RUPTURE_DISTANCE, --rupture_distance RUPTURE_DISTANCE
                        Path to the rupture distance csv file
  -srf SRF_INFO, --srf_info SRF_INFO
                        Path to srf-info file
  -s STATIONS [STATIONS ...], --stations STATIONS [STATIONS ...]
                        List of stations to calculate empiricals for
  -rm MAX_RUPTURE_DISTANCE, --max_rupture_distance MAX_RUPTURE_DISTANCE
                        Only calculate empiricals for stations that are within
                        X distance to rupture
  -i IDENTIFIER, --identifier IDENTIFIER
                        run-name for run
  -c CONFIG, --config CONFIG
                        configuration file to select which model is being used
  -e, --extended_period
                        Indicate the use of extended(100) pSA periods
  -p PERIOD [PERIOD ...], --period PERIOD [PERIOD ...]
                        pSA period(s) separated by a space. eg: 0.02 0.05 0.1.
  -m IM [IM ...], --im IM [IM ...]
                        Intensity measure(s) separated by a space(if more than
                        one). eg: PGV PGA CAV.
```

e.g.
```
python calculate_empirical.py -v non_uniform_whole_nz_with_real_stations-hh400_v18p6.vs30 -r rrups/2012p713691.csv -srf srf/Kelly_HYP01-01_S1244.info -i Kelly_HYP01-01_S1244 output -s CCCC GODS
```

To aggregate the empirical files calculated above:
```
usage: emp_aggregation.py [-h] [-o OUTPUT_DIR] [-i IDENTIFIER] [-r RUPTURE]
                          [-v VERSION]
                          im_files [im_files ...]

positional arguments:
  im_files

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUT_DIR, --output_dir OUTPUT_DIR
                        path to output folder that stores the aggregated
                        measures
  -i IDENTIFIER, --identifier IDENTIFIER
                        run-name for run
  -r RUPTURE, --rupture RUPTURE
                        Please specify the rupture name of the simulation.
                        eg.Albury
  -v VERSION, --version VERSION
                        The version of the simulation. eg.18p4
```
e.g.
```
python emp_aggregation
-o /home/jam335/Documents/Empirical-Engine/Data/output
-i Darfield_Benchmark -r Darfield
/home/jam335/Documents/Empirical-Engine/Data/output/Darfield_AS_16_Ds575.csv
/home/jam335/Documents/Empirical-Engine/Data/output/Darfield_AS_16_Ds595.csv
/home/jam335/Documents/Empirical-Engine/Data/output/Darfield_Br_13_PGA.csv
/home/jam335/Documents/Empirical-Engine/Data/output/Darfield_Br_13_PGV.csv
/home/jam335/Documents/Empirical-Engine/Data/output/Darfield_Br_13_pSA.csv
/home/jam335/Documents/Empirical-Engine/Data/output/Darfield_CB_12_AI.csv
/home/jam335/Documents/Empirical-Engine/Data/output/Darfield_CB_12_CAV.csv
```

To generate all possible aggregations of the empirical files calculated in the first script:
```
usage: aggregate_empirical_im_permutations.py 
[-h] [--version VERSION] simulation_directory

positional arguments:
  simulation_directory  The directory of the event or fault realisation to
                        generate aggregated empirical intensity measure files
                        for

optional arguments:
  -h, --help            show this help message and exit
  --version VERSION, -v VERSION
                        The version of the simulation

```
e.g.
```
python aggregate_empirical_im_permutations.py
/nesi/project/nesi00213/RunFolder/CyberShake/v19p5/Runs/Hossack/Hossack_REL01
--version 19p5
```


CHANGELOG.md
Empirical Engine - Empirical IM calculation

(Based on https://wiki.canterbury.ac.nz/download/attachments/58458136/CodeVersioning_v18p2.pdf?version=1&modificationDate=1519269238437&api=v2 )

## Unreleased
### Fixed
### Changed
### Added

## [21.1.1] - 2021-01-21
### Added
  - Added OpenQuake compatibility layer
  - Added OpenQuake models for ver component
  - Added component selection to command line arguments

## [19.9.3] - 2020-02-11 -
### Added
  - Added RotD100/50 (rotd100_50) and inelastic to elastic spectral displacement ratio (IESDR) empiricals
### Changed
  - Added 16 pSA periods to the default

## [19.9.2] - 2019-12-10 -
### Added
  - Permutation aggregation script to generate all combinations of intensity measure models for a given event or fault from the available intensity measure files
  - Abrahamson A18 model added and verification script created 
### Changed
  - Empirical calculation and aggregation scripts refactored to allow easier access to their functionality

## [19.9.1] - 2019-09-11
### Added
Additional Empirical Models. (and unit tests)

## [18.7.2] - 2018-08-16
### Added
Interpolation for Zhao beyond the period of 5.0s
Config file to select which model is used for each tect_type / im
Empirical IM_csv creation
IM_aggregation (aggregating an individual IM_csv into a larger file containing more than one IM_type)
### Changed
Verification changes (Z1p0 units and rock_site estimation for CB)

## [18.7.1] - 2018-07-13
### Added
Empirical calculations from site (rrup + vs30) and fault for Zhao_06, Bradley_13, Afshari_Stewart_16,
    Campbell_Bozorgina_2010, Campbell_Bozorgina_2012
Output IM into individual csv files
Estimation of parameters unspecified


calculate_empirical.py
#!/usr/bin/env python2
"""Script to calculate IMs for empirical models.

Produces one .csv for each IM containing all specified sites.
"""
import os
import argparse
import h5py
import numpy as np
import pandas as pd

import empirical_factory

from GMM_models.classdef import Site
from GMM_models.classdef import Fault
from GMM_models.classdef import TectType
from GMM_models import classdef

from qcore.utils import setup_dir
from qcore.im import order_im_cols_df

IM_LIST = ['PGA', 'PGV', 'CAV', 'AI', 'Ds575', 'Ds595', 'pSA']
EXT_PERIOD = np.logspace(start=np.log10(0.01), stop=np.log10(10.), num=100,
                         base=10)
PERIOD = [0.02, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.75, 1.0, 2.0, 3.0, 4.0, 5.0,
          7.5, 10.0]

PSA_IM_NAME = 'pSA'
STATION_COL_NAME = 'station'
COMPONENT_COL_NAME = 'component'


def create_fault_parameters(srf_info):
    """Create fault parameters"""
    fault = Fault()
    f = h5py.File(srf_info, 'r')
    attrs = f.attrs
    dip = attrs['dip']
    if np.max(dip) == np.min(dip):
        fault.dip = np.min(dip)
    else:
        print("unexpected dip value")
        exit()
    fault.Mw = np.max(attrs['mag'])
    rake = attrs['rake']
    if np.max(rake) == np.min(rake):
        fault.rake = np.min(rake)
    else:
        print("unexpected rake value")
        exit()
    fault.Mw = attrs['mag']
    if 'dtop' in attrs:
        fault.ztor = np.min(attrs['dtop'])
    else:
        fault.ztor = attrs['hdepth']
    if 'tect_type' in attrs:
        fault.tect_type = TectType[attrs['tect_type']]
    else:
        print("tect_type not found assuming 'ACTIVE_SHALLOW'")
        fault.tect_type = TectType.ACTIVE_SHALLOW
    fault.hdepth = attrs['hdepth']
    return fault


def read_rrup_file(rrup_file):
    """Read rupture(?) file"""
    rrups = dict()
    with open(rrup_file) as f:
        next(f)
        for line in f:
            station, __, __, rrup, rjbs, rx = line.rstrip().split(',')
            rrup = float(rrup)
            rjbs = float(rjbs)
            if rx == "X":
                rx = None
            else:
                rx = float(rx)
            values = (rrup, rjbs, rx)
            rrups[station] = values
    return rrups


def read_vs30_file(vs30_file):
    """Read vs30 file"""
    values = dict()
    with open(vs30_file) as f:
        for line in f:
            station, value = line.split()
            values[station] = float(value)
    return values


def create_site_parameters(rrup_file, vs30_file, stations=None,
                           vs30_default=500, max_distance=None):
    """Create site parameters"""
    rrups = read_rrup_file(rrup_file)
    vs30_values = read_vs30_file(vs30_file)
    sites = list()
    if max_distance is None:
        max_distance = float("inf")
    if stations is None:
        stations = list(rrups.keys())
    for station in stations:
        rrup, rjbs, rx = rrups[station]
        if rrup < max_distance:
            site = Site()
            site.name = station
            if station in vs30_values:
                site.vs30 = vs30_values[station]
            else:
                site.vs30 = vs30_default
            site.Rrup = rrup
            site.Rjb = rjbs
            site.Rx = rx
            sites.append(site)
    return sites


def calculate_empirical():
    """Calculate empirical intensity measures"""
    parser = argparse.ArgumentParser(
        description="Script to calculate IMs for empirical models."
                    "Produces one .csv for each IM containing "
                    "all specified sites.")

    parser.add_argument('--vs30_file', '-v', required=True,
                        help="vs30 file. Default value is 250 if"
                             " station/file not present")
    parser.add_argument('-r', '--rupture_distance', required=True,
                        help="Path to the rupture distance csv file")
    parser.add_argument('-srf', '--srf_info', help="Path to srf-info file",
                        required=True)
    parser.add_argument('--vs30_default', default=classdef.VS30_DEFAULT,
                        help="Sets the default value for the vs30")
    parser.add_argument('-s', '--stations', nargs='+',
                        help="List of stations to calculate empiricals for")
    parser.add_argument('-rm', '--max_rupture_distance',
                        help="Only calculate empiricals for stations "
                             "that are within X distance to rupture")
    parser.add_argument('-i', '--identifier', help="run-name for run")
    parser.add_argument('-c', '--config',
                        help="configuration file to "
                             "select which model is being used")
    parser.add_argument('-e', '--extended_period', action='store_true',
                        help="Indicate the use of extended(100) pSA periods")

    parser.add_argument('-p', '--period', nargs='+', default=PERIOD, type=float,
                        help='pSA period(s) separated by a '
                             'space. eg: 0.02 0.05 0.1.')
    parser.add_argument('-m', '--im', nargs='+', default=IM_LIST,
                        help='Intensity measure(s) separated by a '
                             'space(if more than one). eg: PGV PGA CAV.')

    parser.add_argument('output', help="output directory")
 
    args = parser.parse_args()

    # Fault & Site parameters
    fault = create_fault_parameters(args.srf_info)
    sites = create_site_parameters(args.rupture_distance, args.vs30_file,
                                   stations=args.stations,
                                   vs30_default=args.vs30_default,
                                   max_distance=args.max_rupture_distance)
    
    setup_dir(args.output)

    if args.extended_period:
        period = np.unique(np.append(PERIOD, EXT_PERIOD))
    else:
        period = PERIOD

    model_dict = empirical_factory.read_model_dict(args.config)
    station_names = [site.name for site in
                     sites] if args.stations is None else args.stations
    for im in args.im:
        cur_gmm = empirical_factory.determine_gmm(fault, im, model_dict)

        # File & column names
        cur_filename = '{}_{}_{}.csv'.format(args.identifier, cur_gmm.name, im)
        cur_cols = []
        if im == PSA_IM_NAME:
            for p in period:
                cur_cols.append('{}_{}'.format(im, p))
                cur_cols.append('{}_{}_sigma'.format(im, p))
        else:
            cur_cols.append(im)
            cur_cols.append('{}_sigma'.format(im))

        # Get & save the data
        cur_data = np.zeros((len(sites), len(cur_cols)), dtype=np.float)
        for ix, site in enumerate(sites):
            values = empirical_factory.compute_gmm(fault, site, cur_gmm, im,
                                                   period)
            if im == PSA_IM_NAME:
                cur_data[ix, :] = np.ravel(
                    [[value_tuple[0], value_tuple[1][0]] for value_tuple in
                     values])
            else:
                cur_data[ix, :] = [values[0], values[1][0]]

        df = pd.DataFrame(columns=cur_cols, data=cur_data)
        df[STATION_COL_NAME] = station_names
        df[COMPONENT_COL_NAME] = 'geom'

        # Correct column order
        df = order_im_cols_df(df)

        df.to_csv(os.path.join(args.output, cur_filename), index=False)


if __name__ == '__main__':
    calculate_empirical()
